showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ComponentImpressions.kt

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Passed in the callback of the [visibilityEvents] Modifier. Contains information about the
+ * visibility of a given composable function that uses the visibilityEvents Modifier.
+ */
+data class ShowkaseVisibilityEvent<T>(
+    val key: T,
+    val visibilityPercentage: Float,
+    val boundsInWindow: Rect
+)
+
+/**
+ * Use this modifier to get visibility events for a given Composable. It emits visibility events
+ * when the composable is added to the composition (visible), when its removed from the
+ * composition(invisible), when the activity is backgrounded(visible) and when the activity is
+ * foregrounded(visible). In addition,
+ *
+ * @param key Unique identifier for a given composable function that you use this modifier on
+ * @param onVisibilityChanged Callback that's called when the visibility of a composable function
+ * changes. This event has information about the key of the composable, the percentage of composable
+ * that's visible on the screen and the bounds of the composable.
+ */
+@OptIn(FlowPreview::class)
+fun <T> Modifier.visibilityEvents(
+    key: T,
+    onVisibilityChanged: (ShowkaseVisibilityEvent<T>) -> Unit,
+) = composed {
+    val view = LocalView.current
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val scope = rememberCoroutineScope()
+    val visibilityEvent by rememberUpdatedState(newValue = onVisibilityChanged)
+    val impressionCollector = remember(key) { ImpressionCollector<T>(scope, visibilityEvent) }
+    var visibilityMetadata: VisibilityMetadata? by remember { mutableStateOf(null) }
+
+    registerLifecycleImpressionEvents(key, lifecycle, impressionCollector, visibilityMetadata)
+    registerDisposeImpressionEvents(key, impressionCollector, lifecycle)
+    collectImpressionEvents(key, impressionCollector, visibilityEvent)
+
+    onGloballyPositioned { layoutCoordinates ->
+        visibilityMetadata = layoutCoordinates.visibilityPercentage(view = view)
+        impressionCollector.onLayoutCoordinatesChanged(key, visibilityMetadata!!)
+    }
+}
+
+/**
+ * Used for handling the use case where a composable function is not in composition anymore i.e is
+ * invisible.
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+private fun <T> registerDisposeImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    lifecycle: Lifecycle
+) {
+    DisposableEffect(key, lifecycle) {
+        onDispose {
+            impressionCollector.publishDisposeEvent(
+                key,
+                defaultVisibilityMetadata,
+            )
+        }
+    }
+}
+
+/**
+ * Collects the visibility change events from the internal coroutine flow and invokes the
+ * onVisibilityEvent that the user passed.
+ */
+@SuppressLint("ComposableNaming")
+@FlowPreview
+@Composable
+private fun <T> collectImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    LaunchedEffect(key) {
```

Comment:
maybe the coroutine scope should be keyed instead of this effect, then you don't need to pass the key or worry about the scoping

Reply:
I think LaunchedEffect guarantees that the scope is the "composition" so that shouldn't be a concern as it auto-cancels any ongoing coroutines.